gpt_index/indices/query/query_runner.py

```diff
@@ -237,6 +206,50 @@ def query(
         else:
             return query_obj.query(query_bundle)
 
+    def query(
+        self,
+        query_str_or_bundle: Union[str, QueryBundle],
+        index_id: Optional[str] = None,
+        level: int = 0,
+    ) -> RESPONSE_TYPE:
+        """Run query.
+        
+        NOTE: Relies on mutual recursion between 
+            - QueryRunner.query
+            - QueryRunner.query_transformed
+            - BaseQueryCombiner.run
+        
+        QueryRunner.query resolves the current index struct, 
+            then pass control to BaseQueryCombiner.run.
+        BaseQueryCombiner.run applies query transforms and makes multiple queries on the same index.
+        Each query is passed to QueryRunner.query_transformed for execution.
+            During execution, we recursively calls QueryRunner.query if index is a composable graph.
+        """
+        # Resolve index struct from index_id if necessary
```

Comment:
we need an async version of the outer `query` call 

can prob abstract everything except return statement into a helper function

Reply:
Addressed by https://github.com/jerryjliu/llama_index/pull/929